### PR TITLE
Threading exploration work

### PR DESCRIPTION
**Pull request by gsouquet**
This pull request implements a low fidelity exploration of the threading work. It is based off the reply-chain. To review alongside with matrix-org/matrix-js-sdk#1828
Some changes that aren't pushed yet, are needed on the js-sdk side of things for this exploration work to work properly

Fixes vector-im/element-web#18532

Screen Shot 2021-08-23 at 14 43 05

---
Here's what your changelog entry will look like:

## ✨ Features
 * Threading exploration work ([\#6658](https://github.com/matrix-org/matrix-react-sdk/pull/6658)). Fixes vector-im/element-web#18532.

Preview: https://6128b35e579f98bb372a6366--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: res/css/views/rooms/_EventTile.scss [hunk 1]
```diff
@@ -674,3 +674,60 @@ $hover-select-border: 4px;
...
+
+    .mx_EventTile_roomName {
+        display: none;
+    }
+
+    .mx_EventTile_line {
+        padding-left: 0 !important;
+        order: 10 !important;
+    }
+    .mx_EventTile {
```
Comment (turt2live) [inline res/css/views/rooms/_EventTile.scss:725]: extreme nit: blank line between these, for consistency at least

File: res/css/views/rooms/_EventTile.scss [hunk 2]
```diff
@@ -674,3 +674,60 @@ $hover-select-border: 4px;
...
     }
 }
+
+
+.mx_ThreadInfo:hover {
+    cursor: pointer;
+}
+
+.mx_ThreadView {
+
```
Comment (turt2live) [inline res/css/views/rooms/_EventTile.scss:684]: extreme nit: remove excess blank line, for consistency.

File: res/css/views/rooms/_EventTile.scss [hunk 3]
```diff
@@ -674,3 +674,60 @@ $hover-select-border: 4px;
...
+            padding: 0;
+        }
+    }
+
+    .mx_EventTile_senderDetails {
+        display: flex;
+        align-items: center;
+        gap: 6px;
+        margin-bottom: 6px;
+        a {
```
Comment (turt2live) [inline res/css/views/rooms/_EventTile.scss:700]: extreme nit: blank line for consistency

File: res/css/views/rooms/_EventTile.scss [hunk 4]
```diff
@@ -674,3 +674,60 @@ $hover-select-border: 4px;
...
+}
+
+.mx_ThreadView {
+
+    display: flex;
+    flex-direction: column;
+
+    .mx_ScrollPanel {
+        margin-top: 20px;
+        .mx_RoomView_MessageList {
```
Comment (turt2live) [inline res/css/views/rooms/_EventTile.scss:690]: extreme nit: blank line for consistency

File: res/css/views/rooms/_EventTile.scss [hunk 5]
```diff
@@ -674,3 +674,60 @@ $hover-select-border: 4px;
         margin-right: 0;
     }
 }
+
+
```
Comment (turt2live) [inline res/css/views/rooms/_EventTile.scss:678]: extreme nit: too many blank links (consistency)

File: res/css/views/rooms/_EventTile.scss [hunk 6]
```diff
@@ -674,3 +674,60 @@ $hover-select-border: 4px;
...
+        align-items: center;
+        gap: 6px;
+        margin-bottom: 6px;
+        a {
+            flex: 1;
+            min-width: none;
+            max-width: 100%;
+            display: flex;
+            align-items: center;
+            .mx_SenderProfile {
```
Comment (turt2live) [inline res/css/views/rooms/_EventTile.scss:706]: blank line for consistency

File: res/css/views/rooms/_MessageComposer.scss [hunk 7]
```diff
@@ -340,3 +340,17 @@ limitations under the License.
...
+ */
+
+.mx_MessageComposer.mx_MessageComposer--compact {
+    margin-right: 0;
+    .mx_MessageComposer_wrapper {
+        padding: 0;
+    }
+    .mx_MessageComposer_button:last-child {
+        margin-right: 0;
+    }
```
Comment (turt2live) [inline res/css/views/rooms/_MessageComposer.scss:355]: needs some blank lines for consistency

File: src/components/structures/MessagePanel.tsx [hunk 8]
```diff
@@ -443,6 +448,11 @@ export default class MessagePanel extends React.Component<IProps, IState> {
         // Always show highlighted event
         if (this.props.highlightedEventId === mxEv.getId()) return true;
 
+        const threadingEnabled = SettingsStore.getValue("experimentalThreadSupport");
+        if (threadingEnabled && mxEv.replyEventId && this.props.hideThreadedMessages === true) {
```
Comment (turt2live) [inline src/components/structures/MessagePanel.tsx:452]: if possible, I'd push the `SettingsStore.getValue()` check as late as possible into the conditional for performance reasons.

File: src/components/structures/ThreadPanel.tsx [hunk 9]
```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2016 OpenMarket Ltd
+Copyright 2019 The Matrix.org Foundation C.I.C.
```
Comment (turt2live) [inline src/components/structures/ThreadPanel.tsx:3]: this doesn't seem right

File: src/components/structures/ThreadPanel.tsx [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+interface IState {
+    threads?: Thread[];
+}
+
+/*
+ * Component which shows the filtered file using a TimelinePanel
+ */
+@replaceableComponent("structures.ThreadView")
+class ThreadView extends React.Component<IProps, IState> {
```
Comment (turt2live) [inline src/components/structures/ThreadPanel.tsx:44]: saves having to scroll all the way down to see if it was exported:

```suggestion
export default class ThreadView extends React.Component<IProps, IState> {
```

File: src/components/structures/ThreadView.tsx [hunk 11]
```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2016 OpenMarket Ltd
+Copyright 2019 The Matrix.org Foundation C.I.C.
```
Comment (turt2live) [inline src/components/structures/ThreadView.tsx:3]: wrong copyright

File: src/components/structures/ThreadView.tsx [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+interface IState {
+    replyToEvent?: MatrixEvent;
+    thread?: Thread;
+}
+
+/*
+ * Component which shows the filtered file using a TimelinePanel
+ */
+@replaceableComponent("structures.ThreadView")
+class ThreadView extends React.Component<IProps, IState> {
```
Comment (turt2live) [inline src/components/structures/ThreadView.tsx:53]: `export default`

File: src/components/structures/ThreadView.tsx [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+        }
+    };
+
+    updateThread = (thread?: Thread) => {
+        if (thread) {
+            this.setState({ thread });
+        } else {
+            this.forceUpdate();
+        }
+    };
```
Comment (turt2live) [inline src/components/structures/ThreadView.tsx:109]: These need member access declared (public/private/etc)

File: src/components/structures/ThreadView.tsx [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+
+    public renderEventTile(event: MatrixEvent): JSX.Element {
```
Comment (turt2live) [inline src/components/structures/ThreadView.tsx:111]: hmm, why is this public?

File: src/components/structures/ThreadPanel.tsx [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+
+    public onThreadEventReceived = () => this.updateThreads();
+
+    public updateThreads = (callback?: () => void): void => {
+        this.setState({
+            threads: this.room.getThreads(),
+        }, callback);
+    };
+
+    public renderEventTile(event: MatrixEvent): JSX.Element {
```
Comment (turt2live) [inline src/components/structures/ThreadPanel.tsx:70]: This shouldn't be public

File: src/components/structures/ThreadPanel.tsx [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    public componentWillUnmount(): void {
+        this.room.removeListener("Thread.update", this.onThreadEventReceived);
+        this.room.removeListener("Thread.ready", this.onThreadEventReceived);
+    }
+
+    public onThreadEventReceived = () => this.updateThreads();
+
+    public updateThreads = (callback?: () => void): void => {
```
Comment (turt2live) [inline src/components/structures/ThreadPanel.tsx:64]:
```suggestion
    private updateThreads = (callback?: () => void): void => {
```

File: src/components/structures/ThreadPanel.tsx [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+        this.room.on("Thread.update", this.onThreadEventReceived);
+        this.room.on("Thread.ready", this.onThreadEventReceived);
+    }
+
+    public componentWillUnmount(): void {
+        this.room.removeListener("Thread.update", this.onThreadEventReceived);
+        this.room.removeListener("Thread.ready", this.onThreadEventReceived);
+    }
+
+    public onThreadEventReceived = () => this.updateThreads();
```
Comment (turt2live) [inline src/components/structures/ThreadPanel.tsx:62]:
```suggestion
    private onThreadEventReceived = () => this.updateThreads();
```

File: src/components/views/messages/MessageActionBar.js [hunk 18]
```diff
@@ -170,6 +173,17 @@ export default class MessageActionBar extends React.PureComponent {
         });
     };
 
+    onThreadClick = () => {
```
Comment (turt2live) [inline src/components/views/messages/MessageActionBar.js:176]: public/private/etc

File: src/settings/Settings.tsx [hunk 19]
```diff
@@ -211,6 +211,15 @@ export const SETTINGS: {[setting: string]: ISetting} = {
         supportedLevels: LEVELS_FEATURE,
         default: false,
     },
+    "experimentalThreadSupport": {
```
Comment (turt2live) [inline src/settings/Settings.tsx:214]: Features are always `feature_threads` or similar: prefixed with `feature`, then underscored naming.

**Review by turt2live — COMMENTED** [review]
Generally it looks good. Your tests are angry still, and all of my comments are human linter concerns.

File: res/css/views/rooms/_EventTile.scss (hunk 1 above)
Comment (gsouquet) [reply res/css/views/rooms/_EventTile.scss:725]: I'm happy to change that over here, but I'd be in favour of adding a linting rule to ensure consistency and to not make developers remember all those styling rules

File: res/css/views/rooms/_EventTile.scss (hunk 2 above)
Comment (gsouquet) [reply res/css/views/rooms/_EventTile.scss:684]: There are unfortunately no lint rules for this one, the closest would be [`no-empty-first-line`](https://stylelint.io/user-guide/rules/list/no-empty-first-line), and it doesn't do what we want

File: src/components/views/messages/MessageActionBar.js (hunk 18 above)
Comment (gsouquet) [reply src/components/views/messages/MessageActionBar.js:176]: This one's a JS file, haven't come round to migrating it yet

**Pull request merged**